docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,223 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In a similar way to the real-to-real case, we want to perform a reduction of the full matrix. But the :math:`2 * CW` matrix is complex-valued and so in this case, we will compare to complex scalars.
+
+Due to some constraints on what we can compute efficiently in the numerical case and to keep the number of numerical evaluations to a minimum, we compute the following (albeit surprising) scalar value:
+
+.. math::
+    s := 2 * v^T (real(CW) ur + i * imag(CW) ui)
+
+where :math:`v \in \mathcal{R}^M`, :math:`ur \in \mathcal{R}^N` and :math:`ui \in \mathcal{R}^N`.
+
+Fast complex input numerical evaluation
+"""""""""""""""""""""""""""""""""""""""
+
+We first consider how to compute :math:`s` with a numerical method. To do so, keeping in mind that we're considering :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`, and that :math:`CW = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})`,  we rewrite it as follows:
+
+.. math::
+    \begin{aligned}
+        s &= 2 * v^T (real(CW) ur + i * imag(CW) ui) \\
+          &= 2 * v^T (\frac{1}{2} * \frac{\partial y}{\partial a} ur + i * \frac{1}{2} * \frac{\partial y}{\partial b} ui) \\
+          &= v^T (\frac{\partial y}{\partial a} ur + i * \frac{\partial y}{\partial b} ui) \\
+          &= v^T ((\frac{\partial y}{\partial a} ur) + i * (\frac{\partial y}{\partial b} ui))
+    \end{aligned}
+
+In this formula, we can see that :math:`\frac{\partial y}{\partial a} ur` and :math:`\frac{\partial y}{\partial b} ui` can be evaluated the same way as the fast version for the real-to-real case.
+Once these real-valued quantities have been computed, we can reconstruct the complex vector on the right side and do a dot product with the real-valued :math:`v` vector.
+
+Fast complex input analytical evaluation
+""""""""""""""""""""""""""""""""""""""""
+
+For the analytical case, things are simpler and we rewrite the formula as:
+
+.. math::
+    \begin{aligned}
+        s &= 2 * v^T (real(CW) ur + i * imag(CW) ui) \\
+          &= v^T real(2 * CW) ur + i * v^T imag(2 * CW) ui) \\
+          &= real(v^T (2 * CW)) ur + i * imag(v^T (2 * CW)) ui
+    \end{aligned}
+
+We can thus use the fact that the backward mode AD provides us with an efficient way to compute :math:`v^T (2 * CW)` and then perform a dot product of the real part with :math:`ur` and the imaginary part with :math:`ui` before reconstructing the final complex scalar :math:`s`.
+
+Why not use a complex :math:`u`
+"""""""""""""""""""""""""""""""
+
+At this point, you might be wondering why we did not select a complex :math:`u` and just performed the reduction :math:`2 * v^T CW u'`.
+The problem is that when doing the numerical evaluation, considering :math:`u' = ur' + i ui'`, we would need to compute:
+
+.. math::
+    \begin{aligned}
+        2*CW u' &= (dy/da + i dy/db)(ur' + i ui') \\
```

Comment:
dy/daur' is a little hard to read. Lets also use \frac here?

Reply:
Ho yes I forgot to update that formula from the initial draft, good catch!